### PR TITLE
Failure to get metadata is now retryable error

### PR DESCRIPTION
**Pull request by joereuss12**
Adds and error to config so that if there is a timeout getting the metadata response, we mark as retryable and exit(11). In addition, found that we were marking things as retryable even when the client has no errors in its checked error stack array, so check for that now and return false for retryable if that array is empty.

### CODE REVIEW

**Comment by joereuss12** [comment]
Marking as a draft so that I can add #810 in this

**Comment by joereuss12** [comment]
Can check this works by modifying this line in config/config.go to be a really short timeout:
https://github.com/PelicanPlatform/pelican/blob/1fb0a529676e558a8ffc4da84e0e7f0cb2ab9a0a/config/config.go#L373

as well as modify:
https://github.com/PelicanPlatform/pelican/blob/1fb0a529676e558a8ffc4da84e0e7f0cb2ab9a0a/config/resources/defaults.yaml#L79-L80
to a shorter timeout

It should deal with both of those cases reporting a retryable error (exit code 11).

**Comment by joereuss12** [comment]
Needs to be back-ported to v7.5.x

File: config/config.go
Comment (bbockelm) [inline config/config.go:92]: Rename this symbol to `Metadata`.

File: config/config.go
Comment (bbockelm) [inline config/config.go:147]: This makes the error a single-use: all metadata errors are considered the "same".

Have `MetadataErr` wrap the underlying error.  This allows things like stack traces to propagate through and logic on the original error cause.  That is, add an `Unwrap` error that returns the original cause (see https://pkg.go.dev/errors).

File: config/config.go
Comment (bbockelm) [inline config/config.go:385]: If we make the `MetadataErr` wrappable, we could have a global var, MetadataTimeoutErr:

```
var MetadataTimeoutErr *MetadataErr := &MetadataErr{err: "Timeout when querying metadata"}
```

And then here, we could have
```
var netErr net.Error
if errors.As(err, netError) {
    err = MetadataTimeoutErr.Wrap(err)
} else {
    NewMetadataError(err, "Error occurred when querying for metadata")
}
```

Then, we can simplify for the retryable test:

```
if errors.Is(err, config.MetadataTimeoutErr) {
   ...
}
```

**Review by bbockelm — CHANGES_REQUESTED** [review]
At some point we need to torch the global error registry and use the normal golang error mechanism.  That day is not today, however.

The big request here is to make the metadata err object more flexible, allowing it to wrap errors and just wrapping the timeout in the end, not replacing it.

File: config/config.go
Comment (bbockelm) [inline config/config.go:393]: This doesn't limit to only timeouts.  `*net.Error` appears to have a separate `Timeout` method.

```suggestion
		if errors.As(err, &netErr) && netErr.Timeout() {
```

By taking the underlying error object and turning it to a string, you loose the error hierarchy.  You really want:

```
MetadataTimeoutErr.Wrap(err)
```

and to add an `Unwrap` method.

**Comment by joereuss12** [comment]
@bbockelm is this more of what you are looking for?

File: cmd/plugin.go
Comment (bbockelm) [inline cmd/plugin.go:368]: Hm - this seems brittle.  If we accidentally set retryable to true on success, it'll make the exit code 11, right?

File: config/config.go
Comment (bbockelm) [inline config/config.go:149]: I think the err shouldn't be concatenated here.  Otherwise it's done twice later on.  What does it look like in your testing?

File: config/config.go
Comment (bbockelm) [inline config/config.go:154]: Standard go style is colon-separated, not semicolon.

File: config/config.go
Comment (bbockelm) [inline config/config.go:158]: Missing string comparison for Msg.

File: config/config.go
Comment (bbockelm) [inline config/config.go:93]: Do not export internal variables.

**Review by bbockelm — CHANGES_REQUESTED** [review]
Down to a few minor items.

File: cmd/plugin.go
Comment (joereuss12) [reply cmd/plugin.go:196]: In this case there was an error because we failed to run `config.InitClient` which is what the bool `isConfigErr` indicates. I just deal with the error after parsing our arguments so we can get information on the output file and such. The client should know about the error since we explicitly add the error as well:
https://github.com/PelicanPlatform/pelican/blob/e84f775368c5d981dbfb096ef7351b65de0a610c/cmd/plugin.go#L117-L125

File: cmd/plugin.go
Comment (joereuss12) [reply cmd/plugin.go:368]: This function is used only when we need to bail early (such as with a panic or failing to InitClient). Therefore, we should only ever be failing here so success should never be true. Our exit codes out of this function should be either exit(1) or exit(11). I at least can't think of a case where we'd be successful but end up within this function but correct me if I'm wrong with any of this.

File: config/config.go
Comment (joereuss12) [reply config/config.go:149]: For my hand testing triggering that error it seems to work okay:
```
TransferError = "Attempt #1: Problem initializing the Pelican Client configuration: Error occured when querying for metadata: Get \"https://osg-htc.org/.well-known/pelican-configuration\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
```
Note: it's a timeout error above because I just triggered that error with the timeout instead of the other. No duplication seems to be showing up though.

**Comment by joereuss12** [comment]
@bbockelm check my comments and correct me if I am wrong with any of them or if you still want adjusting. Next round on revisions are in and everything seems to be working how it should.

File: cmd/plugin.go
Comment (bbockelm) [reply cmd/plugin.go:196]: Ah, ok, makes sense.

File: cmd/plugin.go
Comment (bbockelm) [reply cmd/plugin.go:368]: Makes sense, thanks.

**Comment by joereuss12** [comment]
Okay, should be all good now. I think this struggle to make proper errors like this is good since now I will have a good template to move forward with and know what mistakes not to make. Ready for review again.

File: config/config.go
Comment (joereuss12) [reply config/config.go]: Ah yeah, you're right. In the earlier comment I was not sure because `target` is of type `error` but we literally convert it right there lol. Fixed now!

**Comment by joereuss12** [comment]
@bbockelm is this good to merge? I think we need your approval since you requested changes.

**Review by bbockelm — APPROVED** [review]
LGTM.

**Comment by bbockelm** [comment]
@joereuss12 - this is good to go once the merge conflicts are cleaned up.

**Pull request merged**
